js-api-spec/value/calculation.test.ts

```diff
@@ -0,0 +1,302 @@
+// Copyright 2023 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import {
+  Value,
+  SassCalculation,
+  SassNumber,
+  SassString,
+  CalculationOperation,
+  CalculationOperator,
+  CalculationInterpolation,
+} from 'sass';
+import {List} from 'immutable';
+
+import '../utils';
+
+const validCalculationValues = [
+  new SassNumber(1),
+  new SassString('1', {quotes: false}),
+  SassCalculation.calc(new SassNumber(1)),
+  new CalculationOperation('+', new SassNumber(1), new SassNumber(1)),
+  new CalculationInterpolation(''),
+];
+const invalidCalculationValues = [1, '1', new SassString('1', {quotes: true})];
+
+/* When `clamp()` is called with less than three arguments, the list of accepted
+values is much narrower */
+const validClampValues = [
+  new SassString('1', {quotes: false}),
+  new CalculationInterpolation('1'),
+];
+const invalidClampValues = [
+  new SassNumber(1),
+  new SassString('1', {quotes: true}),
+];
+
+const validOperators = ['+', '-', '*', '/'];
+const invalidOperators = ['||', '&&', 'plus', 'minus', ''];
+
+describe('SassCalculation', () => {
+  describe('construction', () => {
+    let calculation: SassCalculation;
+    beforeEach(() => {
+      calculation = SassCalculation.calc(new SassNumber(1));
+    });
+
+    it('is a value', () => {
+      expect(calculation).toBeInstanceOf(Value);
+    });
+
+    it('is a calculation', () => {
+      expect(calculation).toBeInstanceOf(SassCalculation);
+      expect(calculation.assertCalculation()).toBe(calculation);
+    });
+
+    it("isn't any other type", () => {
+      expect(() => calculation.assertBoolean()).toThrow();
+      expect(() => calculation.assertColor()).toThrow();
+      expect(() => calculation.assertFunction()).toThrow();
+      expect(() => calculation.assertMap()).toThrow();
+      expect(calculation.tryMap()).toBe(null);
+      expect(() => calculation.assertNumber()).toThrow();
+      expect(() => calculation.assertString()).toThrow();
+    });
+  });
+
+  describe('calc', () => {
+    it('correctly stores name and arguments', () => {
+      const result = SassCalculation.calc(new SassNumber(1));
+      expect(result.name).toBe('calc');
+      expect(result.arguments).toEqualWithHash(List([new SassNumber(1)]));
+    });
+
+    it('rejects invalid arguments', () => {
+      for (const value of invalidCalculationValues) {
+        // @ts-expect-error
+        expect(() => SassCalculation.calc(value)).toThrow();
```

Comment:
@nex3 Because `invalidCalculationValues` includes non-calculation values like `1` and `"1"`, TS complains about calling `calc()` with them. In this case I'm testing precisely that an exception is thrown at runtime, not just flagged by tsc.

This raises some questions:

- Is it right to have runtime validation of `CalculationValue`? Or should I just rely on TS to flag passing the wrong type of arguments?
- If runtime validation is desired, is this the right way to test for it?

Reply:
As far as the spec is concerned, implementations aren't _required_ to have any particular failure state if callers violate the stated type declarations, so I'd leave off checks for anything that violates the TypeScript types.